File: src/docker_manager.rs
//! Docker Manager / Manager Docker
//! Generates Dockerfiles and docker-compose.yml / Generuje Dockerfile i docker-compose.yml

use std::collections::HashMap;
use std::path::Path;
use tracing::{debug, info, warn};

/// Docker template for different languages / Szablon Docker dla różnych języków
#[derive(Debug, Clone)]
pub struct DockerTemplate {
    pub image: &'static str,
    pub build_cmd: &'static str,
    pub run_cmd: &'static str,
}

pub struct DockerManager {
    hub_dir: String,
    templates: HashMap<String, DockerTemplate>,
}

impl DockerManager {
    pub fn new(hub_dir: String) -> Self {
        let mut templates = HashMap::new();

        templates.insert(
            "python".to_string(),
            DockerTemplate {
                image: "python:3.11-slim",
                build_cmd: "RUN pip install --no-cache-dir -r requirements.txt || echo 'No requirements.txt found'",
                run_cmd: r#"CMD ["python", "main.py"]"#,
            },
        );

        templates.insert(
            "nodejs".to_string(),
            DockerTemplate {
                image: "node:20-slim",
                build_cmd: "RUN npm install || echo 'No package.json found'",
                run_cmd: r#"CMD ["npm", "start"]"#,
            },
        );

        templates.insert(
            "rust".to_string(),
            DockerTemplate {
                image: "rust:1.75-slim",
                build_cmd: r#"RUN cargo build --release"#,
                run_cmd: r#"CMD ["./target/release/app"]"#,
            },
        );

        templates.insert(
            "golang".to_string(),
            DockerTemplate {
                image: "golang:1.21-slim",
                build_cmd: r#"RUN go build -o app ."#,
                run_cmd: r#"CMD ["./app"]"#,
            },
        );

        templates.insert(
            "generic".to_string(),
            DockerTemplate {
                image: "ubuntu:latest",
                build_cmd: "",
                run_cmd: r#"CMD ["sh"]"#,
            },
        );

        Self { hub_dir, templates }
    }

    /// Detect programming language / Wykryj język programowania
    pub fn detect_language(&self, repo_path: &Path) -> String {
        let files: Vec<String> = match std::fs::read_dir(repo_path) {
            Ok(entries) => entries
                .filter_map(|e| e.ok())
                .filter_map(|e| e.file_name().into_string().ok())
                .collect(),
            Err(_) => return "generic".to_string(),
        };

        if files.contains(&"requirements.txt".to_string())
            || files.contains(&"pyproject.toml".to_string())
            || files.contains(&"setup.py".to_string())
        {
            "python"
        } else if files.contains(&"package.json".to_string()) {
            "nodejs"
        } else if files.contains(&"Cargo.toml".to_string()) {
            "rust"
        } else if files.contains(&"go.mod".to_string()) {
            "golang"
        } else {
            "generic"
        }
        .to_string()
    }

    /// Generate Dockerfile / Generuj Dockerfile
    pub fn generate_dockerfile(&self, repo_path: &Path, lang: &str) -> bool {
        let dockerfile_path = repo_path.join("Dockerfile");

        // Skip if Dockerfile already exists / Pomiń jeśli Dockerfile już istnieje
        if dockerfile_path.exists() {
            debug!("Dockerfile already exists at {}", dockerfile_path.display());
            return true;
        }

        let template = match self.templates.get(lang) {
            Some(t) => t,
            None => self.templates.get("generic").unwrap(),
        };

        let content = format!(
            r#"FROM {}
WORKDIR /app
COPY . .
{}
{}
"#,
            template.image, template.build_cmd, template.run_cmd
        );

        match std::fs::write(&dockerfile_path, content) {
            Ok(_) => {
                info!("Generated Dockerfile at {}", dockerfile_path.display());
                true
            }
            Err(e) => {
                warn!("Failed to write Dockerfile: {}", e);
                false
            }
        }
    }

    /// Generate multi-stage Dockerfile for Rust / Generuj multi-stage Dockerfile dla Rust
    pub fn generate_rust_dockerfile(&self, repo_path: &Path) -> bool {
        let dockerfile_path = repo_path.join("Dockerfile");

        if dockerfile_path.exists() {
            return true;
        }

        let content = r#"# Build stage
FROM rust:1.75-slim-bookworm as builder
WORKDIR /app
RUN apt-get update && apt-get install -y pkg-config libssl-dev && rm -rf /var/lib/apt/lists/*
COPY Cargo.toml Cargo.lock ./
RUN mkdir src && echo "fn main() {}" > src/main.rs
RUN cargo build --release
RUN rm -rf src
COPY src ./src
RUN cargo build --release

# Runtime stage
FROM debian:bookworm-slim
WORKDIR /app
RUN apt-get update && apt-get install -y ca-certificates && rm -rf /var/lib/apt/lists/*
COPY --from=builder /app/target/release/app /app/app
RUN useradd -m -u 1000 appuser
USER appuser
ENTRYPOINT ["/app/app"]
"#;

        match std::fs::write(&dockerfile_path, content) {
            Ok(_) => {
                info!("Generated Rust multi-stage Dockerfile");
                true
            }
            Err(e) => {
                warn!("Failed to write Dockerfile: {}", e);
                false
            }
        }
    }

    /// Generate docker-compose.yml / Generuj docker-compose.yml
    pub fn create_docker_compose(&self, cloned_repos: &HashMap<String, String>) -> bool {
        let mut services = String::new();

        for (name, path) in cloned_repos {
            services.push_str(&format!(
                r#"  {}:
    build: {}
    container_name: hub_{}
    restart: always
"#,
                name, path, name
            ));
        }

        let compose_content = format!(
            r#"version: '3.8'

services:
{}"#,
            services
        );

        let compose_path = Path::new(&self.hub_dir).join("docker-compose.yml");

        match std::fs::write(&compose_path, compose_content) {
            Ok(_) => {
                info!("Generated docker-compose.yml at {}", compose_path.display());
                true
            }
            Err(e) => {
                warn!("Failed to write docker-compose.yml: {}", e);
                false
            }
        }
    }

    /// Process all repositories / Przetwórz wszystkie repozytoria
    pub fn process_repositories(&self, repos: &[(String, std::path::PathBuf)]) -> HashMap<String, String> {
        let mut processed = HashMap::new();

        for (name, path) in repos {
            let lang = self.detect_language(path);
            info!("Detected language for {}: {}", name, lang);

            let success = if lang == "rust" {
                self.generate_rust_dockerfile(path)
            } else {
                self.generate_dockerfile(path, &lang)
            };

            if success {
                processed.insert(name.clone(), path.to_string_lossy().to_string());
            }
        }

        // Generate docker-compose if we have multiple repos
        // Generuj docker-compose jeśli mamy wiele repo
        if processed.len() > 1 {
            self.create_docker_compose(&processed);
        }

        processed
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::TempDir;

    #[test]
    fn test_detect_python() {
        let temp_dir = TempDir::new().unwrap();
        fs::write(temp_dir.path().join("requirements.txt"), "flask==2.0").unwrap();

        let manager = DockerManager::new("/tmp".to_string());
        let lang = manager.detect_language(temp_dir.path());
        assert_eq!(lang, "python");
    }

    #[test]
    fn test_detect_rust() {
        let temp_dir = TempDir::new().unwrap();
        fs::write(temp_dir.path().join("Cargo.toml"), "[package]").unwrap();

        let manager = DockerManager::new("/tmp".to_string());
        let lang = manager.detect_language(temp_dir.path());
        assert_eq!(lang, "rust");
    }

    #[test]
    fn test_generate_dockerfile() {
        let temp_dir = TempDir::new().unwrap();
        let manager = DockerManager::new("/tmp".to_string());

        let result = manager.generate_dockerfile(temp_dir.path(), "python");
        assert!(result);
        assert!(temp_dir.path().join("Dockerfile").exists());

        let content = fs::read_to_string(temp_dir.path().join("Dockerfile")).unwrap();
        assert!(content.contains("python:3.11-slim"));
    }
}

File: src/RyzeHub.Application/Hub/DockerManager.cs
using Microsoft.Extensions.Logging;

namespace RyzeHub.Application.Hub;

public sealed record DockerTemplate(string Image, string BuildCommand, string RunCommand);

/// <summary>
/// Generates Dockerfiles and a docker-compose file for repositories cloned into the hub.
/// </summary>
public sealed class DockerManager(ILogger<DockerManager> logger, string hubDirectory)
{
    private static readonly Dictionary<string, DockerTemplate> Templates = new(StringComparer.Ordinal)
    {
        ["dotnet"] = new(
            "mcr.microsoft.com/dotnet/sdk:10.0",
            "RUN dotnet restore && dotnet publish -c Release -o /app/publish",
            "CMD [\"dotnet\", \"/app/publish/app.dll\"]"),
        ["python"] = new(
            "python:3.13-slim",
            "RUN pip install --no-cache-dir -r requirements.txt || echo 'No requirements.txt found'",
            "CMD [\"python\", \"main.py\"]"),
        ["nodejs"] = new(
            "node:22-slim",
            "RUN npm ci || npm install || echo 'No package.json found'",
            "CMD [\"npm\", \"start\"]"),
        ["rust"] = new("rust:1.83-slim", "RUN cargo build --release", "CMD [\"./target/release/app\"]"),
        ["golang"] = new("golang:1.23-alpine", "RUN go build -o app .", "CMD [\"./app\"]"),
        ["java"] = new("eclipse-temurin:21-jdk", "RUN ./mvnw -q -DskipTests package || mvn -q -DskipTests package", "CMD [\"java\", \"-jar\", \"target/app.jar\"]"),
        ["generic"] = new("debian:bookworm-slim", string.Empty, "CMD [\"sh\"]")
    };

    public string DetectLanguage(string repositoryPath)
    {
        string[] files;
        try
        {
            files = [.. Directory.EnumerateFiles(repositoryPath).Select(Path.GetFileName).OfType<string>()];
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            return "generic";
        }

        if (files.Any(file => file.EndsWith(".sln", StringComparison.OrdinalIgnoreCase))
            || Directory.EnumerateFiles(repositoryPath, "*.csproj", SearchOption.AllDirectories).Any())
        {
            return "dotnet";
        }

        if (files.Contains("requirements.txt", StringComparer.Ordinal)
            || files.Contains("pyproject.toml", StringComparer.Ordinal)
            || files.Contains("setup.py", StringComparer.Ordinal))
        {
            return "python";
        }

        if (files.Contains("package.json", StringComparer.Ordinal))
        {
            return "nodejs";
        }

        if (files.Contains("Cargo.toml", StringComparer.Ordinal))
        {
            return "rust";
        }

        if (files.Contains("go.mod", StringComparer.Ordinal))
        {
            return "golang";
        }

        return files.Contains("pom.xml", StringComparer.Ordinal) ? "java" : "generic";
    }

    public bool GenerateDockerfile(string repositoryPath, string language)
    {
        var dockerfilePath = Path.Combine(repositoryPath, "Dockerfile");

        if (File.Exists(dockerfilePath))
        {
            logger.LogDebug("Dockerfile already exists at {DockerfilePath}", dockerfilePath);
            return true;
        }

        var template = Templates.GetValueOrDefault(language, Templates["generic"]);
        var content = $"""
            FROM {template.Image}
            WORKDIR /app
            COPY . .
            {template.BuildCommand}
            {template.RunCommand}

            """;

        try
        {
            File.WriteAllText(dockerfilePath, content);
            logger.LogInformation("Generated Dockerfile at {DockerfilePath}", dockerfilePath);
            return true;
        }
        catch (IOException exception)
        {
            logger.LogWarning(exception, "Failed to write Dockerfile");
            return false;
        }
    }

    public bool GenerateDotnetDockerfile(string repositoryPath)
    {
        var dockerfilePath = Path.Combine(repositoryPath, "Dockerfile");
        if (File.Exists(dockerfilePath))
        {
            return true;
        }

        const string Content = """
            # Build stage
            FROM mcr.microsoft.com/dotnet/sdk:10.0 AS build
            WORKDIR /src
            COPY . .
            RUN dotnet restore
            RUN dotnet publish -c Release -o /app/publish --no-restore /p:UseAppHost=false

            # Runtime stage
            FROM mcr.microsoft.com/dotnet/aspnet:10.0 AS runtime
            WORKDIR /app
            COPY --from=build /app/publish .
            RUN adduser --disabled-password --gecos "" --uid 1000 appuser
            USER appuser
            ENTRYPOINT ["dotnet", "app.dll"]

            """;

        try
        {
            File.WriteAllText(dockerfilePath, Content);
            logger.LogInformation("Generated .NET multi-stage Dockerfile");
            return true;
        }
        catch (IOException exception)
        {
            logger.LogWarning(exception, "Failed to write Dockerfile");
            return false;
        }
    }

    public bool CreateDockerCompose(IReadOnlyDictionary<string, string> clonedRepositories)
    {
        var services = string.Concat(clonedRepositories.Select(pair => $"""
              {pair.Key.ToLowerInvariant().Replace('.', '-')}:
                build: {pair.Value}
                container_name: hub_{pair.Key.ToLowerInvariant().Replace('.', '-')}
                restart: unless-stopped

            """));

        var composePath = Path.Combine(hubDirectory, "docker-compose.yml");

        try
        {
            Directory.CreateDirectory(hubDirectory);
            File.WriteAllText(composePath, $"services:\n{services}");
            logger.LogInformation("Generated docker-compose.yml at {ComposePath}", composePath);
            return true;
        }
        catch (IOException exception)
        {
            logger.LogWarning(exception, "Failed to write docker-compose.yml");
            return false;
        }
    }

    public IReadOnlyDictionary<string, string> ProcessRepositories(IReadOnlyList<(string Name, string Path)> repositories)
    {
        var processed = new Dictionary<string, string>(StringComparer.Ordinal);

        foreach (var (name, path) in repositories)
        {
            var language = DetectLanguage(path);
            logger.LogInformation("Detected language for {Repository}: {Language}", name, language);

            var success = language == "dotnet"
                ? GenerateDotnetDockerfile(path)
                : GenerateDockerfile(path, language);

            if (success)
            {
                processed[name] = path;
            }
        }

        if (processed.Count > 1)
        {
            CreateDockerCompose(processed);
        }

        return processed;
    }
}
